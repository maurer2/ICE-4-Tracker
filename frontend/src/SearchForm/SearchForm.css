.form {
  margin-bottom: 1rem;
}

.form-row {
  position: relative;
  display: flex;
}

.loader {
  position: absolute;
  right: 0;
  top: 0;
  bottom: 0;
  height: 100%;
  padding: 0.5rem;
  width: auto;
}
