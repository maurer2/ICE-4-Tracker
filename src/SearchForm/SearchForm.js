import React, { Component } from 'react';

import styles from './SearchForm.module.css';

import SearchField from '../SearchField/SearchField';
import SuggestionList from '../SuggestionList/SuggestionList';
import Spinner from '../Spinner/Spinner';

class SearchForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      showSuggestions: false,
      newTrainNumber: '',
    };

    this.handleSearchFieldFocusEvent = this.handleSearchFieldFocusEvent.bind(this);
    this.handleKeyboardEvents = this.handleKeyboardEvents.bind(this);
    this.handleClickEvent = this.handleClickEvent.bind(this);
  }

  handleSearchFieldFocusEvent(searchFieldHasFocus) {
    this.setState({showSuggestions: searchFieldHasFocus})
  }

  handleKeyboardEvents(key) {
    const allowedKeys = ['ArrowUp', 'ArrowDown', 'Enter'];

    if (!allowedKeys.includes(key)) {
      return;
    }

    if (key === 'Enter') {
      const { newTrainNumber } = this.state;

      this.props.cbUpdateActiveTrain(newTrainNumber);
      //this.searchField.blur();
      // workaround
      document.activeElement.blur();

      return;
    }

    if (key === 'ArrowDown') {
      this.updateSelectedTrainNumber(true);
      
      return;
    }

    this.updateSelectedTrainNumber(false);
  }

  handleClickEvent(trainNumber) {
    this.setState({
      newTrainNumber: trainNumber,
      showSuggestions: false,
    }, () => {
      this.props.cbUpdateActiveTrain(trainNumber);
    });
  }

  updateSelectedTrainNumber(goDown) {
    if (this.props.trainNumbers.length === 0) {
      return;
    }

    const { newTrainNumber } = this.state;
    const { trainNumbers } = this.props;

    const firstTrainNumber = trainNumbers[0];
    const lastTrainNumber = trainNumbers[trainNumbers.length - 1];
    if (goDown) {
      // last entry
      if (newTrainNumber === lastTrainNumber) {
        return;
      }

      const nextIndex = trainNumbers.indexOf(newTrainNumber) + 1;

      this.setState({ newTrainNumber: trainNumbers[nextIndex] });
    } else {
      // first entry
      if (newTrainNumber === firstTrainNumber) {
        return;
      }

      const prevIndex = trainNumbers.indexOf(newTrainNumber) - 1;

      this.setState({ newTrainNumber: trainNumbers[prevIndex] });
    }
  }

  render() {
    const { showSuggestions, newTrainNumber } = this.state;
    const { showLoader, trainNumbers, activeTrain } = this.props;

    return (
      <form
        action="/"
        className={ styles.form }
        method="post"
        autoComplete="off"
        noValidate="novalidate"
        onSubmit={ (event) => event.preventDefault() }
      >
        <h2 className={ styles.title }>
          <label htmlFor="search-field">
            Nummern&shy;suche
          </label>
        </h2>
        <div className={ styles['form-row'] }>
          { showLoader && (
            <div className={ styles['spinner-overlay'] }>
              <Spinner />
            </div>
          )}
          <SearchField
            cbFieldFocus={ this.handleSearchFieldFocusEvent }
            cbKeyboardEvent={ this.handleKeyboardEvents }
            value={ activeTrain }
            id="search-field"
            name="search-field"
            ref={ c => this.searchField = c }
          />
        </div>
        <div className={ styles.formRow }>
          <SuggestionList
            showSuggestions={ showSuggestions }
            trainNumbers={ trainNumbers }
            selectedTrainNumber={ newTrainNumber }
            cbHandleClickEvent={ this.handleClickEvent }
          />
        </div>
      </form>
    )
  }
}

export default SearchForm;
