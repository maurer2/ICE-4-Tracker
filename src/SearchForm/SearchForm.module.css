.form {
  margin-bottom: 2rem;
}

.title {
  margin: 0 0 0.5rem 0;
  text-align: left;
  font-weight: normal;
  font-size: 1.25rem;
  color: white;
}

.form-row {
  position: relative;
  display: flex;
}

.spinner-overlay {
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  padding: 0.5rem;
}
