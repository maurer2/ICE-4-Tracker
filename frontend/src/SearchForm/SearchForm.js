import React, { Component } from 'react';
import './SearchForm.css';
import { ReactComponent as Loader } from './loader.svg';

import SearchField from '../SearchField/SearchField';
import SuggestionList from '../SuggestionList/SuggestionList';

class SearchForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      showSuggestions: false,
      inputValue: this.props.activeTrain,
      newTrainNumber: '',
    };

    this.handleSearchFieldFocusEvent = this.handleSearchFieldFocusEvent.bind(this);
    this.handleKeyboardEvents = this.handleKeyboardEvents.bind(this);
    this.handleClickEvents = this.handleClickEvents.bind(this);
  }

  handleSearchFieldFocusEvent(searchFieldHasFocus) {
    this.setState({
      showSuggestions: true,
    })
  }

  handleKeyboardEvents(key) {
    const allowedKeys = ['ArrowUp', 'ArrowDown', 'Enter'];

    if (!allowedKeys.includes(key)) {
      return;
    }

    if (key === 'Enter') {
      const newTrain = this.state.newTrainNumber;

      this.props.cbTrainChangeEvent(newTrain);

      return;
    }

    this.updateSelectedTrainNumber(key === 'ArrowDown');
  }

  handleClickEvents(trainNumber) {
    this.setState({
      newTrainNumber: trainNumber
    }, () => {
      this.props.cbTrainChangeEvent(trainNumber);
    });
  }

  updateSelectedTrainNumber(goDown) {
    if (this.props.trainNumbers.length === 0) {
      return;
    }

    const selectedTrainNumber = this.state.newTrainNumber;
    const firstTrainNumber = this.props.trainNumbers[0];
    const lastTrainNumber = this.props.trainNumbers[this.props.trainNumbers.length - 1];

    if (goDown) {
      if (selectedTrainNumber === lastTrainNumber) {
        return;
      }

      const nextIndex = this.props.trainNumbers.indexOf(selectedTrainNumber) + 1;

      this.setState({
        newTrainNumber: this.props.trainNumbers[nextIndex]
      })
    } else {
      if (selectedTrainNumber === firstTrainNumber) {
        return;
      }

      const prevIndex = this.props.trainNumbers.indexOf(selectedTrainNumber) - 1;

      this.setState({
        newTrainNumber: this.props.trainNumbers[prevIndex]
      })
    }
  }

  render() {
    return (
      <form action="/" className="form" method="post" autoComplete="off" noValidate="novalidate"
        onSubmit={ (e) => e.preventDefault() }>
        <h2 className="subtitle">
          Nummern&shy;suche
        </h2>
        <div className="form-row">
          { this.props.showLoader && <Loader className="loader"></Loader> }
          <SearchField cbFieldFocus={ this.handleSearchFieldFocusEvent } cbKeyboardEvent={ this.handleKeyboardEvents } />
        </div>
        <div className="form-row">
          <SuggestionList showSuggestions={ this.state.showSuggestions } trainNumbers={ this.props.trainNumbers }
            selectedTrainNumber={ this.state.newTrainNumber } cbHandleClickEvents={ this.handleClickEvents } />
        </div>
      </form>
    )
  }
}

export default SearchForm;
