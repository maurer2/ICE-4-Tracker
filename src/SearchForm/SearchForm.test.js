import React from 'react';
import { shallow } from 'enzyme';

import SearchForm from './SearchForm';

describe('SearchForm', () => {
  const mockUpdateTrain = jest.fn();

  const wrapper = shallow(
    <SearchForm
      trainNumbers={['500', '555']}
      showLoader={false}
      activeTrain="500"
      cbUpdateActiveTrain={mockUpdateTrain}
    />
  );

  test('SearchForm should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
