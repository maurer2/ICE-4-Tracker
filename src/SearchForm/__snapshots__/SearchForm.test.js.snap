// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm SearchForm should match snapshot 1`] = `
<form
  action="/"
  autoComplete="off"
  className="form"
  method="post"
  noValidate="novalidate"
  onSubmit={[Function]}
>
  <h2
    className="title"
  >
    <label
      htmlFor="search-field"
    >
      NummernÂ­suche
    </label>
  </h2>
  <div
    className="form-row"
  >
    <SearchField
      cbFieldFocus={[Function]}
      cbKeyboardEvent={[Function]}
      id="search-field"
      name="search-field"
    />
  </div>
  <div
    className="formRow"
  >
    <SuggestionList
      cbHandleClickEvent={[Function]}
      selectedTrainNumber=""
      showSuggestions={false}
      trainNumbers={
        Array [
          "500",
          "555",
        ]
      }
    />
  </div>
</form>
`;
